pkg/spiffs/fs/spiffs_fs.c

```diff
@@ -0,0 +1,486 @@
+    fs_desc->config.log_block_size = dev->page_size * dev->pages_per_sector;
+    fs_desc->config.log_page_size = dev->page_size;
+    fs_desc->config.phys_addr = 0;
+    fs_desc->config.phys_erase_block = dev->page_size * dev->pages_per_sector;
+#endif
+
+    mtd_init(dev);
+
+    s32_t ret = SPIFFS_mount(&fs_desc->fs,
+                             &fs_desc->config,
+                             fs_desc->work,
+                             fs_desc->fd_space,
+                             SPIFFS_FS_FD_SPACE_SIZE,
+#if SPIFFS_CACHE == 1
+                             fs_desc->cache,
+                             SPIFFS_FS_CACHE_SIZE,
+#else
+                             NULL,
+                             0,
+#endif
+                             NULL);
+
+    if (ret != 0) {
+        DEBUG("spiffs: mount: ret %" PRId32 "\n", ret);
+        switch (ret) {
+        case SPIFFS_ERR_NOT_A_FS:
+            DEBUG("spiffs: mount: formatting fs\n");
+            ret = SPIFFS_format(&fs_desc->fs);
+            DEBUG("spiffs: mount: format ret %" PRId32 "\n", ret);
+            if (ret < 0) {
+                return spiffs_err_to_errno(ret);
+            }
+            ret = SPIFFS_mount(&fs_desc->fs,
+                               &fs_desc->config,
+                               fs_desc->work,
+                               fs_desc->fd_space,
+                               SPIFFS_FS_FD_SPACE_SIZE,
+#if SPIFFS_CACHE == 1
+                               fs_desc->cache,
+                               SPIFFS_FS_CACHE_SIZE,
+#else
+                               NULL,
+                               0,
+#endif
+                               NULL);
+            DEBUG("spiffs: mount: ret %" PRId32 "\n", ret);
+            break;
+        }
+    }
+
+    return spiffs_err_to_errno(ret);
+}
+
+static int _umount(vfs_mount_t *mountp)
+{
+    spiffs_desc_t *fs_desc = mountp->private_data;
+
+    SPIFFS_unmount(&fs_desc->fs);
+
+    return 0;
+}
+
+static int _unlink(vfs_mount_t *mountp, const char *name)
+{
+    spiffs_desc_t *fs_desc = mountp->private_data;
+
+    return spiffs_err_to_errno(SPIFFS_remove(&fs_desc->fs, name));
+}
+
+static int _rename(vfs_mount_t *mountp, const char *from_path, const char *to_path)
+{
+    spiffs_desc_t *fs_desc = mountp->private_data;
+
+    return spiffs_err_to_errno(SPIFFS_rename(&fs_desc->fs, from_path, to_path));
+}
+
+static int _open(vfs_file_t *filp, const char *name, int flags, mode_t mode, const char *abs_path)
+{
+    spiffs_desc_t *fs_desc = filp->mp->private_data;
+    (void) abs_path;
+    DEBUG("spiffs: open: private_data = %p\n", filp->mp->private_data);
+
+    spiffs_flags s_flags = SPIFFS_O_RDONLY;
+    if (flags & O_APPEND) {
+        s_flags |= SPIFFS_O_APPEND;
+    }
+    if (flags & O_TRUNC) {
+        s_flags |= SPIFFS_O_TRUNC;
+    }
+    if (flags & O_CREAT) {
+        s_flags |= SPIFFS_O_CREAT;
+    }
+    if (flags & O_WRONLY) {
+        s_flags |= SPIFFS_O_WRONLY;
+    }
+    if (flags & O_RDWR) {
```

Comment:
There's also a SPIFFS_O_RDONLY which should be set 
```c
if ((flags & O_RDONLY) == O_RDONLY) {
    s_flags |= SPIFFS_O_RDONLY;
}
```
(must be checked for equality because O_RDONLY is defined as 0 by default in newlib)